src/Nixfmt/Pretty.hs

```diff
@@ -0,0 +1,141 @@
+
+splitN :: NodeType -> [NixAST] -> Maybe ([NixAST], NixAST, [NixAST])
+splitN n = split $ \case (Node n' _) -> n == n'
+                         _           -> False
+
+thenLine :: NixAST -> Doc
+thenLine node = pretty node <> space
+
+instance Pretty NixAST where
+    pretty (Leaf l Nothing) = pretty l
+    pretty (Leaf l (Just comment)) = pretty l <> text (" #" <> comment) <> hardline
+    pretty (Trivia t) = pretty t
+
+    pretty (Node File children) = pretty children <> hardline
+
+    pretty (Node List children0) =
+        let Just (leading, brackOpen, children1) = splitT TBrackOpen children0
+            Just (items, brackClose, trailing) = splitT TBrackClose children1
+        in pretty leading <>
+           group (thenLine brackOpen <>
+                  nest 2 (hcat (map thenLine items)) <>
+                  pretty brackClose) <>
+           pretty trailing
+
+    pretty (Node Set children0) =
+        let Just (leading, braceOpen, children1) = splitT TBraceOpen children0
+            Just (bindings, braceClose, trailing) = splitT TBraceClose children1
+        in pretty leading <>
+           group (thenLine braceOpen <>
+                  nest 2 (hcat (map thenLine bindings)) <>
+                  thenLine braceClose) <>
+           pretty trailing
+
+    pretty (Node Assignment children0) =
+        let Just (leading, assignToken, children1) = splitT TAssign children0
+            Just (expression, semicolon, trailing) = splitT TSemicolon children1
+        in pretty leading <> text " " <> pretty assignToken <> softline <>
+           pretty expression <> pretty semicolon <> pretty trailing
+
+    pretty (Node With children0) =
+        let Just (leading, withToken, children1) = splitT TWith children0
+            Just (expr1, semicolon, expr2) = splitT TSemicolon children1
+        in pretty leading <> pretty withToken <> text " " <>
+           pretty expr1 <> pretty semicolon <> softline <> pretty expr2
+
+    pretty (Node Let children0) =
+        let Just (leading, letToken, children1) = splitT TLet children0
+            Just (bindings, inToken, expression) = splitT TIn children1
+        in pretty leading <> space <> thenLine letToken <>
+           nest 2 (hcat (map (group . thenLine) bindings)) <>
+           emptyline <> pretty inToken <> text " " <> pretty expression
+
+    pretty (Node Abstraction children0) =
+        let Just (parameter, colon, expression) = splitT TColon children0
+        in group $ pretty parameter <> pretty colon <> space <> pretty expression
+
+    pretty (Node ContextParameter children0) = hcat $ map pretty children0
+
+    pretty (Node SetParameter children0) =
+        let Just (leading, braceOpen, children1) = splitT TBraceOpen children0
+            Just (params, braceClose, trailing) = splitT TBraceClose children1
+            go xs = case splitT TComma xs of
+                         Just (before, comma, after)
+                                 -> nest 2 (pretty before) <> break <>
```

Comment:
> `break`

It would be lovely if, instead of hiding standard well-known definitions, you imported and used your custom combinators qualified.